### PR TITLE
Initial work on chat backend

### LINKED ISSUE #111: Add web sockets to support chat system
## Feature
https://faye.jcoglan.com/node.html
https://github.com/uWebSockets/uWebSockets
One channel + Direct messages
Hash of two people’s names to create private channel
Authentication
* Each message has to be checked on server before publishing to clients
* For DMs make sure they are allowed in
* Rate limiting - Rocket League style
* Use full names, or just first name + first letter name
* Mongo pagination
Need
* Endpoint for active users in chat
* Within users, need to know whether they are actually active
* ServiceWorker for background notifications

### PR DESCRIPTION
**Pull request by antoniomika**
Closes #111

### PR Description
Initial work on chat backend
### Context?
Still a WIP

### CODE REVIEW

**Comment by antoniomika** [comment]
What is currently implemented:
- [x] Authentication for user (cookie or token based)
- [x] Rate limit checking (configurable)
- [x] User checking (server verifies all messages)
- [x] Channels/user lists, and updates when new user joins
- [x] Admin omniscient view
- [x] Profile fetching
- [x] Demo page at `/chat`

File: server/db/model/User.js [hunk 1]
```diff
@@ -272,39 +273,57 @@ schema.methods.changePassword = function(password) {
 };
 
 schema.methods.sendVerificationEmail = function() {
-    Email.sendEmailTemplate(
-        config.confirmation_email_template,
-        {
-            confirmation_url: config.host +
+    if (config.development) {
```
Comment (connorkrupp) [inline server/db/model/User.js]: oh thats so nice

File: server/socketio/chat.js [hunk 2]
```diff
@@ -0,0 +1,177 @@
+var Responses = require('../responses/socketio/index.js');
+
+module.exports = function(io) {
+    setInterval(function() {
+        //console.log('Channel Members', getChannelMembers(io, '#general'));
```
Comment (connorkrupp) [inline server/socketio/chat.js]: don't need this im guessing?

File: server/views/chat.html [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+    <script src="https://code.jquery.com/jquery-3.2.1.min.js" type="text/javascript"></script>
+    <script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>
+
+    <script src="/socket.io/socket.io.js"></script>
+    <script>
+
+        var profileData = null,
+            socket = null;
+
+        //connect();
```
Comment (connorkrupp) [inline server/views/chat.html]: probs extra

File: server/views/chat.html [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+
+        function connect(token) {
+            socket = io.connect('http://localhost:3000', {
+                reconnection: false
+            });
+
+            socket.on('authenticate', function (data) {
+                console.log('Authenticate', data);
+                if (!data) {
+                    authenticate(socket, token || 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJlbWFpbCI6ImFobWlrYUB1bWljaC5lZHUiLCJpYXQiOjE0OTYzMjgzNTEsImV4cCI6MTQ5NzUzNzk1MX0.SwyWgu8-bgBvG7oaf07myrv8IVHQh6lQKeaL1n2q2wY')
```
Comment (connorkrupp) [inline server/views/chat.html]: lol this is quite the default token

**Comment by connorkrupp** [comment]
So is the chat.html for testing or what should we do with it

File: server/socketio/chat.js (hunk 2 above)
Comment (antoniomika) [inline server/socketio/chat.js]: Nooope

**Comment by antoniomika** [comment]
The `chat.html` or `/chat` route is just for demo purposes. It gives Rob an idea of the logic and it helps me with making sure features work as intended. I'll be using a lot of similar logic when I help job implement the chat ui

File: server/views/chat.html (hunk 3 above)
Comment (antoniomika) [inline server/views/chat.html]: This file is gonna get removed eventually

File: server/views/chat.html (hunk 4 above)
Comment (antoniomika) [inline server/views/chat.html]: Same as above

**Comment by roball24** [comment]
approved to get the ball rolling

**Pull request merged**